main.go

```diff
@@ -95,75 +158,128 @@ func (pl promLogger) Println(v ...interface{}) {
+		}
+	}
+	return nil
+}
+
+func (c collectorsConfigMap) asSlice() []string {
+	var cols []string
+	for col := range c {
+		cols = append(cols, col)
+	}
+	return cols
+}
+
+func (c collectorsConfigMap) isEmpty() bool {
+	return len(c.asSlice()) == 0
+}
+
+func (c *collectorsConfigMap) Type() string {
 	return "string"
 }
 
 type options struct {
-	inCluster     bool
-	apiserver     string
-	kubeconfig    string
-	help          bool
-	port          int
-	host          string
-	telemetryPort int
-	telemetryHost string
-	collectors    collectorSet
-	namespace     string
-	version       bool
+	inCluster                bool
+	apiserver                string
+	kubeconfig               string
+	help                     bool
+	port                     int
+	host                     string
+	telemetryPort            int
+	telemetryHost            string
+	collectors               collectorsMap
+	collectorsConfig         collectorsConfigMap
+	listRegisteredCollectors bool
+	namespace                string
+	version                  bool
 }
 
 func main() {
-	options := &options{collectors: make(collectorSet)}
+	collectorsConfig := collectorsConfigMap(map[string]string(supportedCollectors))
+
+	options := &options{collectors: make(collectorsMap), collectorsConfig: make(collectorsConfigMap)}
 	flags := pflag.NewFlagSet("", pflag.ExitOnError)
 	// add glog flags
 	flags.AddGoFlagSet(flag.CommandLine)
 	flags.Lookup("logtostderr").Value.Set("true")
 	flags.Lookup("logtostderr").DefValue = "true"
 	flags.Lookup("logtostderr").NoOptDefVal = "true"
 	flags.BoolVar(&options.inCluster, "in-cluster", true, `If true, use the built in kubernetes cluster for creating the client`)
+	flags.BoolVar(&options.listRegisteredCollectors, "list-registered-collectors", false, `If true, list registered collectors with corresponding supported group versions`)
 	flags.StringVar(&options.apiserver, "apiserver", "", `The URL of the apiserver to use as a master`)
 	flags.StringVar(&options.kubeconfig, "kubeconfig", "", "Absolute path to the kubeconfig file")
 	flags.BoolVarP(&options.help, "help", "h", false, "Print help text")
 	flags.IntVar(&options.port, "port", 80, `Port to expose metrics on.`)
 	flags.StringVar(&options.host, "host", "0.0.0.0", `Host to expose metrics on.`)
 	flags.IntVar(&options.telemetryPort, "telemetry-port", 81, `Port to expose kube-state-metrics self metrics on.`)
 	flags.StringVar(&options.telemetryHost, "telemetry-host", "0.0.0.0", `Host to expose kube-state-metrics self metrics on.`)
-	flags.Var(&options.collectors, "collectors", fmt.Sprintf("Comma-separated list of collectors to be enabled. Defaults to %q", &defaultCollectors))
+	flags.Var(&options.collectors, "collectors", fmt.Sprintf("Comma-separated list of collectors to be enabled. Defaults to %q", &supportedCollectors))
+	flags.Var(&options.collectorsConfig, "collectors-config", fmt.Sprintf("Comma-separated list of collectors group version to be used. Defaults to %q", &collectorsConfig))
```

Comment:
I feel this gives the notion to the user, that multiple versions are actually supported, which we don't, so I would just limit this PR to automatically enabling the collectors that are available/discovered in the API.

Reply:
> so I would just limit this PR to automatically enabling the collectors that are available/discovered in the API

IIUC, you mean that for now, there is only one group version support for each collectors and we need to add all available/discovered collectors in the API to kube-state-metrics.

Yes. For now, there is only one group version for each collector. I prepare to add a second group version to cronjob as it has been discussed for a long time in https://github.com/kubernetes/kube-state-metrics/issues/337 and other issues. And after that we suggest we cut a release to let end users to give this feature a try and we can collect feedback earlier.

As for other additional group version support for different collectors, they should be added on-demand when only end users request for them.

Reply:
Yeah, I get that we could add new collectors for different versions, but that looks like a whole lot of code duplication right now. I'm not sure it's that big of a problem, as starting with 1.9 there is no collector that is still in alpha, feels like we're over-complicating it.

Reply:
If we choose to use this implementation, it will make code duplication indeed. Maybe we can do some abstract and refactor to extract common code and support for custom with arguments.